clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -1082,6 +1097,7 @@ Node selectReadReplica(TopicPartition partition, Node leaderReplica, long curren
         Map<Node, FetchSessionHandler.FetchRequestData> reqs = new LinkedHashMap<>();
         for (Map.Entry<Node, FetchSessionHandler.Builder> entry : fetchable.entrySet()) {
             reqs.put(entry.getKey(), entry.getValue().build());
+            this.nodesWithPendingFetchRequests.add(entry.getKey().id());
```

Comment:
I wonder if it would be better to make this call after `client.send` returns successfully in `sendFetches`? 

Another safety we might add is to remove from `nodesWithPendingFetchRequests` above if `client.isUnavailable` returns true?

Reply:
@hachikuji  Thanks for the review. Moved the call after `client.send`.

I wasn't sure about removing for `client.isUnavailable`. I think we guarantee that the listener `onSuccess` or `onFailure` is invoked in all cases if `send` suceeds. So that should be sufficient? I was a bit concerned that an additional remove could in theory mean that a listener invoked later from heartbeat thread could potentially remove the subsequent send. Dont think it can happen in practice though.

Reply:
Yeah, I think that should be fine. I just wasn't sure how much I wanted to trust the request completion handling logic. I guess it would break a lot of expectations elsewhere if it were broken though, so probably no additional harm from relying on it here.